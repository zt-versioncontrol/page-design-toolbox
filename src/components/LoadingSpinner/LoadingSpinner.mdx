import { Meta, ArgTypes } from '@storybook/blocks'
import LoadingSpinner from '@/components/LoadingSpinner/LoadingSpinner.vue'
import * as Stories from './LoadingSpinner.stories.ts'
import {
  ComponentSummary,
  CustomTypes,
  ComponentDependency,
  Examples,
  ExampleGroup
} from '/.storybook/documents/DocBlocks'

<Meta title="LoadingSpinner Document" of={Stories} component={LoadingSpinner} />

<ComponentSummary of={Stories} />

<ArgTypes of={Stories} />

<Examples>
  <ExampleGroup title="Default Loading Spinner" stories={[Stories.DefaultSpinner]} />
  <ExampleGroup
    title="Style Variants"
    stories={[Stories.StyleVariant1, Stories.StyleVariant2, Stories.StyleVariant3]}
  />
</Examples>
