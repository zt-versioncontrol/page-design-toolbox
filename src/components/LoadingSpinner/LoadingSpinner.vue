<script setup lang="ts">
import { LOADING_ANIMATION_CYCLE } from '@/standards/Time'
import { IN_SCRIPT_ICON } from '@/standards/Dimensions'
import { MAIN_TEXT, DISABLED } from '@/standards/Colors'

// TODO: add optional loading message to display below the spinner

export interface Props {
  /**
   * size of spinner, both height and width. May be any valid CSS size value
   */
  size?: string
  /**
   * spinner track color. May be any valid CSS color
   */
  trackColor?: string
  /**
   * spinner color. May be any valid CSS color
   */
  spinnerColor?: string
}

const props = withDefaults(defineProps<Props>(), {
  size: IN_SCRIPT_ICON,
  trackColor: DISABLED,
  spinnerColor: MAIN_TEXT
})
</script>

<template>
  <div class="spinner"></div>
</template>

<style scoped>
@keyframes spin {
  0% {
    transform: rotate(0deg);
  }

  100% {
    transform: rotate(360deg);
  }
}

.spinner {
  /* Font size is used to set border width relative to size using em */
  font-size: v-bind(size);
  border: 0.1em solid v-bind(trackColor);
  border-top: 0.1em solid v-bind(spinnerColor);
  border-radius: 10rem;
  height: v-bind(size);
  width: v-bind(size);
  animation: spin;
  animation-duration: v-bind('LOADING_ANIMATION_CYCLE + "s"');
  animation-iteration-count: infinite;
  animation-timing-function: linear;
}
</style>
