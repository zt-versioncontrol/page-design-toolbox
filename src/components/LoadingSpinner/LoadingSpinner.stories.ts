import type { Meta, StoryObj } from '@storybook/vue3'
import LoadingSpinner from '@/components/LoadingSpinner/LoadingSpinner.vue'
import { type Documentation } from '@/../.storybook/documents/DocUtilities'
import * as componentTags from '@/../.storybook/documents/ComponentTags'
import * as componentBadges from '@/../.storybook/documents/ComponentBadges'

const meta: Meta<typeof LoadingSpinner> = {
  title: 'Interactivity/LoadingSpinner',
  component: LoadingSpinner,
  excludeStories: ['documentation']
}

export default meta

type Story = StoryObj<typeof LoadingSpinner>

export const documentation: Documentation = {
  title: 'LoadingSpinner',
  summary:
    'Loading Spinner is used to indicate pending operation. Mainly used while waiting for data from a server',
  tags: [componentTags.ICON],
  badges: [componentBadges.RELEASED]
  // TODO: add testing guide
}

export const DefaultSpinner: Story = {
  args: {}
}

export const StyleVariant1: Story = {
  args: {
    spinnerColor: 'red',
    trackColor: '#ff000055'
  }
}

export const StyleVariant2: Story = {
  args: {
    spinnerColor: 'green',
    trackColor: '#00ffff55',
    size: '4rem'
  }
}

export const StyleVariant3: Story = {
  args: {
    spinnerColor: 'blue',
    trackColor: '#ffffff00',
    size: '5rem'
  }
}
