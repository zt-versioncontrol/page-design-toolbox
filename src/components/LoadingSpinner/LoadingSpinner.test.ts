import { mount } from '@vue/test-utils'
import { describe, it, expect } from 'vitest'
import LoadingSpinner from '@/components/LoadingSpinner/LoadingSpinner.vue'

describe('Default component', () => {
  it('should be rendered', () => {
    const wrapper = mount(LoadingSpinner, {
      attachTo: document.body
    })

    expect(wrapper.isVisible()).toBe(true)
  })

  it('should match snapshot', () => {
    const wrapper = mount(LoadingSpinner, {})

    expect(wrapper.element).toBeTruthy()
    expect(wrapper.element).toMatchSnapshot()
  })
})
