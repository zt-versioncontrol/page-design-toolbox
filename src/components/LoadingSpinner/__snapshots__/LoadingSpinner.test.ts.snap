// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`Default component > should match snapshot 1`] = `
<div
  class="spinner"
  data-v-bbf2ec71=""
/>
`;
